### PR TITLE
Refactor au-core-medreq-03 (FHIR-46498, FHIR-48362)

### PR DESCRIPTION
**Pull request by JayMurdoch**
[FHIR-46498](https://jira.hl7.org/browse/FHIR-46498): medreq-03 allows non-precise date if Data Absent Reason is present
[FHIR-48362](https://jira.hl7.org/browse/FHIR-48362): correct to require min length of 10

* Refactored au-core-medreq-03 for precise date or DAR, increased min length to 10

### CODE REVIEW

**Comment by dt-r** [comment]
The PR should include both technical corrections on this invariant.

File: input/pagecontent/changes.md [hunk 1]
```diff
@@ -46,7 +46,7 @@ This change log documents the significant updates and resolutions implemented fr
...
   - added Must Support to MedicationRequest.reasonReference [FHIR-45090](https://jira.hl7.org/browse/FHIR-45090)
   - support for AMT codes and PBS Item Codes in MedicationRequest.medicationCodeableConcept changed to define slicing in AU Core (no longer inherited from AU Base as per [FHIR-44821](https://jira.hl7.org/browse/FHIR-44821)). Slicing discriminator changed from slicing by value:system to slicing by value set and the binding strength has been corrected to required. [FHIR-46391](https://jira.hl7.org/browse/FHIR-46391)
-  - updated invariant au-core-medreq-03 to allow for a Data Absent Reason extension in order to meet AU Core Missing Data and Suppressed Data requirements, and changed the length check from >= 10 to >= 8 to match required precision to the day [FHIR-46417](https://jira.hl7.org/browse/FHIR-46417)
+  - updated invariant au-core-medreq-03 to allow for a Data Absent Reason extension in order to meet AU Core Missing Data and Suppressed Data requirements, and corrected the FHIRPath expression [FHIR-46417](https://jira.hl7.org/browse/FHIR-46417), [FHIR-46498](https://jira.hl7.org/browse/FHIR-46498), [FHIR-48362](https://jira.hl7.org/browse/FHIR-48362)
```
Comment (dt-r) [inline input/pagecontent/changes.md:49]: remove the change log entry. This is just QA and isn't a substantive change.

File: input/resources/au-core-medicationrequest.xml [hunk 2]
```diff
@@ -248,7 +248,7 @@
         <key value="au-core-medreq-03"/>
         <severity value="error"/>
         <human value="Date shall be precise to the day or, if not available, the Data Absent Reason extension shall be present"/>
-        <expression value="(toString().length() &gt;= 8) xor extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()"/>
+        <expression value="($this is DateTime implies $this.toString().length() &gt;= 10) xor extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()"/>
```
Comment (dt-r) [inline input/resources/au-core-medicationrequest.xml:251]: This still allows a non-precise date & a data absent reason, i.e.:
<authoredOn value="2020-07">
    <extension url="http://hl7.org/fhir/StructureDefinition/data-absent-reason">
      <valueCode value="masked"/>
    </extension>
</authoredOn>

Does not trigger the failure case for this invariant.

File: input/pagecontent/changes.md (hunk 1 above)
Comment (JayMurdoch) [reply input/pagecontent/changes.md:49]: Removed.

File: input/resources/au-core-medicationrequest.xml (hunk 2 above)
Comment (JayMurdoch) [reply input/resources/au-core-medicationrequest.xml:251]: Amended, thanks.

Comment (dt-r) [reply input/resources/au-core-medicationrequest.xml]: No need for this complexity. Two simpler options are:

`$this is DateTime implies $this.toString().length() &gt;= 10 xor extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()`

or 

`$this.where($this.toString().length() &gt;= 10).exists() xor extension('
[http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()`

**Review by dt-r — CHANGES_REQUESTED** [review]
See continued conversation

File: input/resources/au-core-medicationrequest.xml (hunk 2 above)
Comment (JayMurdoch) [reply input/resources/au-core-medicationrequest.xml]: Thanks for the suggestions. I have tested both and have included results below - neither was successful. I have kept the complex invariant.

$this is DateTime implies $this.toString().length() &gt;= 10 xor extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()
|Test|Outcome|
| --- | ---| 
|Valid DAR|pass|
|Valid date (to day)|pass|
|Valid date (to day) + DAR|fail|
|Valid date (to seconds)|pass|
|Valid date (to seconds) + DAR|fail|
|Invalid date (to month)|fail| 
|Invalid date (to month) + DAR|**pass**| 

$this.where($this.toString().length() &gt;= 10).exists() xor extension(' [http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()
|Test|Outcome|
| --- | ---| 
|Valid DAR|**fail**|
|Valid date (to day)|pass|
|Valid date (to day) + DAR|**pass**|
|Valid date (to seconds)|pass|
|Valid date (to seconds) + DAR|**pass**|
|Invalid date (to month)|fail| 
|Invalid date (to month) + DAR|fail|

Comment (dbojicic) [reply input/resources/au-core-medicationrequest.xml]: @JayMurdoch, could you please test this expression? I think it works and is the simplest one I could get to work (I think). 
 
`<expression value="(value.exists() and value.toString().length() &gt;= 10 and extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists().not()) or (value.empty() and extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists())"/>`

**Review by dbojicic — CHANGES_REQUESTED** [review]
Changes requested.

File: input/resources/au-core-medicationrequest.xml (hunk 2 above)
Comment (JayMurdoch) [reply input/resources/au-core-medicationrequest.xml]: Test results below - all behaving as expected. Thank you!

|Test|Outcome|
| --- | ---| 
|Valid DAR|pass|
|Valid date (to day)|pass|
|Valid date (to day) + DAR|fail|
|Valid date (to seconds)|pass|
|Valid date (to seconds) + DAR|fail|
|Invalid date (to month)|fail| 
|Invalid date (to month) + DAR|fail|

File: input/resources/au-core-medicationrequest.xml [hunk 3]
```diff
@@ -248,7 +248,7 @@
         <key value="au-core-medreq-03"/>
         <severity value="error"/>
         <human value="Date shall be precise to the day or, if not available, the Data Absent Reason extension shall be present"/>
-        <expression value="(toString().length() &gt;= 8) xor extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()"/>
+        <expression value="(value.exists() and value.toString().length() &gt;= 10 and extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists().not()) or (value.empty() and extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists())"/>
         <xpath value="(f:matches(f:effectiveDateTime, /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z)) and not(exists(f:extension[@url='http://hl7.org/fhir/StructureDefinition/data-absent-reason']))) or (not(f:matches(f:effectiveDateTime, /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))) and exists(f:extension[@url='http://hl7.org/fhir/StructureDefinition/data-absent-reason']))"/>
```
Comment (dbojicic) [inline input/resources/au-core-medicationrequest.xml:252]: please update xpath to match the updated expression.

**Review by dbojicic — CHANGES_REQUESTED** [review]
Please see inline comments

File: input/resources/au-core-medicationrequest.xml (hunk 3 above)
Comment (dbojicic) [reply input/resources/au-core-medicationrequest.xml:252]: I am not convinced this is the solution. Could not get it to work with FHIRPath Lab either. 
I've raised a question in Zulip: [MedicationRequest.authoredOn allow precise date or DAR ext](https://chat.fhir.org/#narrow/channel/179266-fhirpath/topic/MedicationRequest.2EauthoredOn.20allow.20precise.20date.20or.20DAR.20ext); @JayMurdoch, could you please follow up if there is no response?

**Comment by dbojicic** [comment]
@JayMurdoch 
based on the feedback in Zulip, here is a refined expression that seem to be working: 

`<expression value =" ($this.hasValue().not() and extension('
[http://hl7.org/fhir/StructureDefinition/data-absent-reason').value.exists())](http://hl7.org/fhir/StructureDefinition/data-absent-reason%27).value.exists()))
or ($this.hasValue() and $this.toString().length() >= 10 and extension('
[http://hl7.org/fhir/StructureDefinition/data-absent-reason').value.exists().not())](http://hl7.org/fhir/StructureDefinition/data-absent-reason%27).value.exists().not()))"/>`

@heathfrankel provided a simplified version of this:
`<expression value =" $this.hasValue() implies (extension('
[http://hl7.org/fhir/StructureDefinition/data-absent-reason').value.exists().not()](http://hl7.org/fhir/StructureDefinition/data-absent-reason%27).value.exists().not())
and $this.toString().length() >= 10)"/>`

Initial test in Fhirpath-lab using HAPI validator passed, see 
:test_tube: [Test with FHIRPath-Lab](https://fhirpath-lab.com/FhirPath?parameters=N4IgpgHgDgTmDO8CWB7AdiAXCAJAFwAsl4A6AgQ3gDVyAbAVzAAoBKAAiQFspakE2mkPGDTJ0TAOQE8eKJgD08grQDsJFDADm8gGZEY8gMp4Y9AMZ56cACJgdSNEjyo08gCbk85ALTkARvAieN5wlOgSLCQAbnSMJJDEePCsJGgoeKxs5GhubPhEpHgoxjAOmim0IpqEmQB8ALxsAIwADCwgADQgZujCEHhYIOT0hBpgbgDyGF1VDmCDAFbkMQIAEgCCAAoAku1dwjCcDii0KJoAnoZgMFHXg9Ky8Ary8ACOp2h6SDAhACwk5AAXlYwAB3MABJwIVJgPC6fSdEAxUr+SrwLAAbQAul04PAUFYzGAAFL4jDYYAAHRAeIJMCJABVzlAwNTMNSALLjJBmTwuABKYFejHgeGpHWpSDcbOpyE4MVFfLQ3igAXF1M4sPIbKpIFgKHslTZGOpDzkimUag0mgB9Hh33kPTgRhM5ksNjsDicLnkw28TrA3k1bh5fPQcGFCDFICxAF8JSBFZZ4DKhhYkLd1SAHMI0NH2SANG5rlng6HnOgAMIoYuosDVtBEqD53U9ENoTTG3Xwc6isCcVNm56q0iaFBRW2OmtgeRQgedalt1lYakADhUAGYmgBRLMh+A8cjnVNyhVeZxoNi-FpsTiaNhePyVPAdNgblrU2M46l9fOyrhnkqV43neD51i+b43o+z4piA8ayvQfgLGAFg6tScA6NcIhEqmmx8kEShHggvgomgaDLgm+6HseK4gOspHkWwqxEbBCZSkESD2CWWC6ngzLLpgrY1mUXayr2wjzgWQ6KAcRxpKcFxkKo6haPI1bFoY4n9vIUQAEzeC0uktBuWZLqmABy2x7sQ1EWWGaB0Gw9COJGHA5BmUr0I57F5pxfAwJ+34gL+qbbKsVlwQmPZ9pJpoyOa8iiPElQWDA6A8gQYB0IQMJ4JOUpKEgs5EPITQkB+C5IrEAlri0xkAGwtKuq4bhu9W-C1ACcKifvBQwjAQYyTGgqZGUZBlNFmAD6wyjHAw1oeA-QiGII2YrqVi0IO8XPJaKnaF8BglG6IK2PYjgVq4HheL4ARBCEWVklmMQMPW06ps5ADWaSgmtX59RGIoHItGFYY2NUgJsMDkOml3XPImgDaUbiaIGSwwPN1kHrQR6ptYMAkGwxLkBj4xsAA4kjUqo5+lE2StYCCpGoqLWg9CcH41wTDogosp48DrLQpzgtKWAAKwJsK2TOHxi0vYwbLvgmzlOKm0GwrT1KQCyFjjIY9BQDw5zWFY9ly9VistMrF2ph4va9ZRKDwOQqPbKIrrpugonBUIqYMuQn1gGw6BB+reBsB4SC0DRCbOHJnY8ehYB8y21I6ID2E0ZgTQJiypQ1myOfUnnqBuAAqjbtGi7GAMEsIi1tiJ61iTF22yM88BpMGJAODoKAvKhlVmbRun1e174tPVWO2bREzQ7QbBpSMy5frHvuz-PmsgG4TtgOsOT8p4AkYrqO+BAAil5vmy4nVWvWy17W6rtF3r1cZxnBQA)

Could you please test both expressions?

**Comment by JayMurdoch** [comment]
@dbojicic Test results - behaved as expected
> `<expression value ="($this.hasValue().not() and extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').value.exists()) or ($this.hasValue() and $this.toString().length() >= 10 and extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').value.exists().not())"/>`
|Test|Outcome|
| --- | ---| 
|Valid DAR|pass|
|Valid date (to day)|pass|
|Valid date (to day) + DAR|fail|
|Valid date (to seconds)|pass|
|Valid date (to seconds) + DAR|fail|
|Invalid date (to month)|fail| 
|Invalid date (to month) + DAR|fail| 

Test results - behaved as expected
> `<expression value ="$this.hasValue() implies (extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').value.exists().not() and $this.toString().length() >= 10)"/>`
|Test|Outcome|
| --- | ---| 
|Valid DAR|pass|
|Valid date (to day)|pass|
|Valid date (to day) + DAR|fail|
|Valid date (to seconds)|pass|
|Valid date (to seconds) + DAR|fail|
|Invalid date (to month)|fail| 
|Invalid date (to month) + DAR|fail|

File: input/pagecontent/changes.md (hunk 1 above)
Comment (dbojicic) [reply input/pagecontent/changes.md]: Change log entry is still present.

File: input/resources/au-core-medicationrequest.xml [hunk 4]
```diff
@@ -248,8 +248,8 @@
         <key value="au-core-medreq-01"/>
         <severity value="error"/>
         <human value="Date shall be precise to the day or, if not available, the Data Absent Reason extension shall be present"/>
-        <expression value="(toString().length() &gt;= 8) xor extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').exists()"/>
-        <xpath value="(f:matches(f:effectiveDateTime, /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z)) and not(exists(f:extension[@url='http://hl7.org/fhir/StructureDefinition/data-absent-reason']))) or (not(f:matches(f:effectiveDateTime, /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))) and exists(f:extension[@url='http://hl7.org/fhir/StructureDefinition/data-absent-reason']))"/>
+        <expression value="$this.hasValue() implies (extension('http://hl7.org/fhir/StructureDefinition/data-absent-reason').value.exists().not() and $this.toString().length() >= 10)"/>
+        <xpath value="(f:matches(f:effectiveDateTime, /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z)) and not(exists(f:extension[@url='http://hl7.org/fhir/StructureDefinition/data-absent-reason'])))"/>
```
Comment (dbojicic) [inline input/resources/au-core-medicationrequest.xml:252]: Please update the xpath to match.

**Review by dbojicic — CHANGES_REQUESTED** [review]
Thanks @JayMurdoch , just a few minor changes.

File: input/pagecontent/changes.md (hunk 1 above)
Comment (dbojicic) [reply input/pagecontent/changes.md]: Thanks @Jaymee, I misread the changes, apologies for that. 
This entry still needs a change to update the invariant key to au-core-medreq-03

Comment (JayMurdoch) [reply input/pagecontent/changes.md]: No worries. Amended, thanks.

File: input/resources/au-core-medicationrequest.xml (hunk 4 above)
Comment (JayMurdoch) [reply input/resources/au-core-medicationrequest.xml]: Xpath updated.

File: input/pagecontent/changes.md [hunk 5]
```diff
@@ -63,7 +63,7 @@ This change log documents the significant updates and resolutions implemented fr
   - removed Must Support from MedicationRequest.substitution, MedicationRequest.substitution.allowed[x] [FHIR-45088](https://jira.hl7.org/browse/FHIR-45088)
   - added Must Support to MedicationRequest.reasonReference [FHIR-45090](https://jira.hl7.org/browse/FHIR-45090)
   - support for AMT codes and PBS Item Codes in MedicationRequest.medicationCodeableConcept changed to define slicing in AU Core (no longer inherited from AU Base as per [FHIR-44821](https://jira.hl7.org/browse/FHIR-44821)). Slicing discriminator changed from slicing by value:system to slicing by value set and the binding strength has been corrected to required. [FHIR-46391](https://jira.hl7.org/browse/FHIR-46391)
-  - updated invariant au-core-medreq-03 to allow for a Data Absent Reason extension in order to meet AU Core Missing Data and Suppressed Data requirements, and changed the length check from >= 10 to >= 8 to match required precision to the day [FHIR-46417](https://jira.hl7.org/browse/FHIR-46417)
+  - updated invariant au-core-medreq-01 to allow for a Data Absent Reason extension in order to meet AU Core Missing Data and Suppressed Data requirements [FHIR-46417](https://jira.hl7.org/browse/FHIR-46417)
```
Comment (dbojicic) [inline input/pagecontent/changes.md:66]: Sorry Jaymee, just noticed another thing. This change should not be in Release 1.0.0-ballot change log. Please remove it from there and add as an entry in the change log for Release 1.0.0 (from version 1.0.0-ballot to 1.0.0).

**Review by dbojicic — CHANGES_REQUESTED** [review]
One minor change

File: input/pagecontent/changes.md (hunk 5 above)
Comment (JayMurdoch) [reply input/pagecontent/changes.md:66]: Amended, thanks.

**Review by dbojicic — APPROVED** [review]

**Pull request merged**